### ISSUE TITLE
[Enhancement] - Matrix-style (or formula driven) testing using molecule

### ISSUE THREAD

**Issue by elliotweiser** [metacloud/molecule#674]
# Issue Type

- Feature Request

# Molecule and Ansible details

```
ansible 2.2.0.0
molecule 1.17.0
```

- Molecule installation method: pip
- Ansible installation method: pip

# Desired Behaviour

As a user of Molecule, I would like the ability to generate a test matrix (or a list of test runs) from within a single `molecule.yml`. Currently, this is only achievable with multiple `molecule.yml` files or doing surgery on a pre-existing one. The benefit of such a feature is more _complete_ testing of an Ansible role. Different role variables can change the final state of the machine under test, which should be accounted during the test run. 

Below is a (somewhat abbreviated) list of the `molecule test` cycle and what is required in order to execute it.

1. Destroy previous state (`molecule destroy`) -- Nothing needed.
1. Set initial state (`molecule create`) -- I need a base image/box.
1. Converge to final state (`molecule converge`) -- I need a playbook
1. Repeated execution results in same final state (`molecule idempotence`) -- Same playbook
1. Verify the final state is correct (`molecule verify`) -- I need a set of tests.

The formula is: Given a vagrant box or docker image, a playbook, and a set of testinfra/serverspec tests, does convergence, idempotence, and verification all pass successfully?

To summarize, I want the ability to supply a (generative) list of that metadata (box, playbook, and tests) to define a matrix/list of test executions.

**Comment by retr0h** [comment]
You could probably accomplish that with [tox](http://molecule.readthedocs.io/en/stable-1.17/usage.html#tox).  This is mostly outside the scope of molecule, but I could see molecule being leveraged to do what you desire.

[Tox factors](http://tox.readthedocs.io/en/latest/config.html#factors-and-factor-conditional-settings) should give you what you want.

**Comment by retr0h** [comment]
BTW - molecule v2 will handle this.  It is scenario based, but quite a bit different than molecule v1, nor does it presently contain all the v1 functionality.

You can play with the v2 branch if you like.  Docs are [up to date](http://molecule.readthedocs.io/en/v2/).

v2 has a scenario provision, converge, test, teardown cycle.

**Comment by elliotweiser** [comment]
I'm somewhat familiar with `tox`. Unless the factor conditions can be "injected" into the molecule.yml, which is not something I was able to figure out, I don't think it's the solution I'm looking for.

I'll check out the v2 branch when I get a chance and play around with it. I'm curious to see the direction this going. Is there currently a backlog for the v2 feature-set? If so, does it exist in any form that is visible to me (such that I may potentially contribute)?

**Comment by retr0h** [comment]
> I'm somewhat familiar with tox. Unless the factor conditions can be "injected" into the molecule.yml, which is not something I was able to figure out, I don't think it's the solution I'm looking for.

An approach one person took #596.

> I'll check out the v2 branch when I get a chance and play around with it. I'm curious to see the direction this going. Is there currently a backlog for the v2 feature-set? If so, does it exist in any form that is visible to me (such that I may potentially contribute)?

Yeah, I'm bad like that.  I haven't really created any issues for features, since I'm primarily the only person that works on it (and in my free time).

Check out the docs and the branch.  I would be super happy to have help.  Molecule v2 leverages ansible to manage instances, so figuring out how to integrate with OpenStack would be nice.  I know thats what you guys currently use.

**Comment by elliotweiser** [comment]
Unfortunately, I've already gone to similar lengths with vagrant boxes, but it's still not the degree of test coverage I'm looking for:

https://github.com/elliotweiser/ansible-osx-command-line-tools/blob/master/molecule.yml#L12-L22

I can still only test with the one playbook and one set of testinfra.

We definitely do use the openstack driver at Comcast, but my preference would be to start with the vagrant feature-set. I'll have some time to look into that this weekend. I'll check out the docs and see what I can do to get started.

Thanks for the assistance and I appreciate the opportunity to contribute to this project.

**Comment by retr0h** [comment]
@elliotweiser sounds good.  Let me know if v2 molecule would accomplish what you are trying to do.  I don't believe we would be able to incorporate what you are trying to do in v1, given the way v1 is designed.  However, I would be interested to know if v2 solves the problem, so that we could close this issue.

**Comment by retr0h** [comment]
Closing, please re-open if you are able to provide feedback.  I'm pretty sure v2 will provide the ability to do what you wish.

**Comment by elliotweiser** [comment]
@retr0h Confirming that v2 addresses the need discussed here. Scenarios FTW 👍

**Comment by retr0h** [comment]
@elliotweiser awesome!
